### PR TITLE
Auto pivot intake absolute encoders

### PR DESCRIPTION
**Pull request by impossiblerocket**
WORKS- press A for down, press A again for up...etc.
B stops intake pivot from turning the motors at all

### CODE REVIEW

File: TitanRobotics2024/src/main/java/frc/robot/Subsystem/IntakePivot.java [hunk 1]
```diff
@@ -0,0 +1,149 @@
...
+import edu.wpi.first.math.controller.ArmFeedforward;
+import edu.wpi.first.math.controller.PIDController;
+import edu.wpi.first.math.controller.ProfiledPIDController;
+import edu.wpi.first.wpilibj.smartdashboard.SmartDashboard;
+import frc.robot.Data.ButtonMap;
+import frc.robot.Data.PortMap;
+import edu.wpi.first.math.trajectory.TrapezoidProfile;
+import edu.wpi.first.math.trajectory.TrapezoidProfile.State;
+import frc.robot.Subsystem.ModifiedMotors;
+import frc.robot.Subsystem.ModifiedEncoders;
```
Comment (thebuilder6) [inline TitanRobotics2024/src/main/java/frc/robot/Subsystem/IntakePivot.java]:
```suggestion
import edu.wpi.first.math.controller.ArmFeedforward;
import edu.wpi.first.math.controller.ProfiledPIDController;
import edu.wpi.first.wpilibj.smartdashboard.SmartDashboard;
import frc.robot.Data.PortMap;
import edu.wpi.first.math.trajectory.TrapezoidProfile;
```
remove unneeded imports

File: TitanRobotics2024/src/main/java/frc/robot/Subsystem/IntakePivot.java [hunk 2]
```diff
@@ -0,0 +1,149 @@
...
+    private double kSVolts = 0.0;
+    private double kGVolts = 0.0;
+    private double kVVolts = 0.0;
+    private double kAVolts = 0.0;
+    private double currentPosition;
+
+    private double maxVelocity;
+    private double maxAcceleration;
+    private double encoderDistancePerRotation = 360.0;
+    private double encoderPositionOffset = 0.0;
```
Comment (thebuilder6) [inline TitanRobotics2024/src/main/java/frc/robot/Subsystem/IntakePivot.java:30]:
```suggestion
```
unused? or should it be the 120 on line 143

File: TitanRobotics2024/src/main/java/frc/robot/Subsystem/IntakePivot.java [hunk 3]
```diff
@@ -0,0 +1,149 @@
...
+                setDisabled(true);
+                break;
+            case "up":
+                goal = 385.0;
+                setDisabled(false);
+                break;
+            case "down":
+                goal = 212.0;
+                setDisabled(false);
+                break;
```
Comment (thebuilder6) [inline TitanRobotics2024/src/main/java/frc/robot/Subsystem/IntakePivot.java]: Make Labled Constants for numbers

File: TitanRobotics2024/src/main/java/frc/robot/Subsystem/IntakePivot.java [hunk 4]
```diff
@@ -0,0 +1,149 @@
...
+        }
+        if (pivotEncoder != null)
+        {
+            IntakeStateProcess();
+        }
+        currentPosition = pivotEncoder.getAbsolutePosition();
+        if (currentPosition <= 120)
+        {
+            currentPosition = (currentPosition + 360);
+        }
```
Comment (thebuilder6) [inline TitanRobotics2024/src/main/java/frc/robot/Subsystem/IntakePivot.java]:
```suggestion
        if (pivotEncoder != null)
        {
            currentPosition = pivotEncoder.getAbsolutePosition();
            if (currentPosition <= 120)
            {
                currentPosition += 360;
            }
            IntakeStateProcess();
        } 
        if (!disabled)
        {
            control();
        }
        else
        {
            pivotMotor.setVoltage(0.0);
        }
```
Reorder to read encoder, process state, then process controller, and move motor. full process in one cycle

**Review by thebuilder6 — CHANGES_REQUESTED** [review]

File: TitanRobotics2024/src/main/java/frc/robot/Subsystem/IntakePivot.java (hunk 2 above)
Comment (impossiblerocket) [reply TitanRobotics2024/src/main/java/frc/robot/Subsystem/IntakePivot.java:30]: reject

File: TitanRobotics2024/src/main/java/frc/robot/Subsystem/Control.java [hunk 5]
```diff
@@ -94,18 +97,21 @@ private void limelightControl()
             {
                 targeting.setTarget("Amp");
                 turn = targeting.aprilTaglockOn();
+                System.out.println("Locking on to Amp");
```
Comment (dalittlet) [inline TitanRobotics2024/src/main/java/frc/robot/Subsystem/Control.java:100]: Remove system.out.println lines

File: TitanRobotics2024/src/main/java/frc/robot/Subsystem/Control.java [hunk 6]
```diff
@@ -94,18 +97,21 @@ private void limelightControl()
...
                 targeting.setTarget("Source");
                 turn = targeting.aprilTaglockOn();
+                System.out.println("Locking on to Source");
             }
 
             if (driverController.getButton(ButtonMap.XboxB))
             {
                 targeting.setTarget("Stage");
                 turn = targeting.aprilTaglockOn();
+                System.out.println("Locking on to Stage");
```
Comment (dalittlet) [inline TitanRobotics2024/src/main/java/frc/robot/Subsystem/Control.java]:
```suggestion
```

File: TitanRobotics2024/src/main/java/frc/robot/Subsystem/Control.java [hunk 7]
```diff
@@ -94,18 +97,21 @@ private void limelightControl()
...
                 targeting.setTarget("Amp");
                 turn = targeting.aprilTaglockOn();
+                System.out.println("Locking on to Amp");
             }
 
             if (driverController.getButton(ButtonMap.XboxA))
             {
                 targeting.setTarget("Source");
                 turn = targeting.aprilTaglockOn();
+                System.out.println("Locking on to Source");
```
Comment (dalittlet) [inline TitanRobotics2024/src/main/java/frc/robot/Subsystem/Control.java]:
```suggestion
```

File: TitanRobotics2024/src/main/java/frc/robot/Subsystem/Control.java [hunk 8]
```diff
@@ -119,6 +125,7 @@ private void limelightControl()
             if (driverController.getButton(ButtonMap.XboxY))
             {
                 turn = targeting.otherLockOn();
+                System.out.println("Locking On to Note");
```
Comment (dalittlet) [inline TitanRobotics2024/src/main/java/frc/robot/Subsystem/Control.java]:
```suggestion
```

File: TitanRobotics2024/src/main/java/frc/robot/Subsystem/IntakePivot.java [hunk 9]
```diff
@@ -0,0 +1,147 @@
...
+            instance = new IntakePivot();
+        }
+        return instance;
+    }
+
+    public IntakePivot()
+    {
+        pivotMotor = new ModifiedMotors(PortMap.INTAKEMOTORPIVOT.portNumber, "CANSparkMax");
+        pivotEncoder = new ModifiedEncoders(PortMap.INTAKEPIVOTENCODER.portNumber, encoderPositionOffset, "DutyCycleEncoder");
+        if (pivotMotor == null)
```
Comment (dalittlet) [inline TitanRobotics2024/src/main/java/frc/robot/Subsystem/IntakePivot.java]: Remove this, this does nothing

File: TitanRobotics2024/src/main/java/frc/robot/Subsystem/ModifiedEncoders.java [hunk 10]
```diff
@@ -21,23 +23,27 @@ public ModifiedEncoders(int channelA, int channelB, String encodertype)
...
         switch (encodertype)
         {
             case "CANSparkMaxEncoder":
                 break;
+            case "DutyCycleEncoder": //absolute encoder
+                dutyCycleEncoder = initializeDutyCycle(portNumber, positionOffset);
+                break;
             default:
-                System.err.println("Error: encoders not activated");
+                System.err.println("Error: single channel encoders not activated");
```
Comment (dalittlet) [inline TitanRobotics2024/src/main/java/frc/robot/Subsystem/ModifiedEncoders.java:45]:
```suggestion
```

File: TitanRobotics2024/src/main/java/frc/robot/Subsystem/ModifiedEncoders.java [hunk 11]
```diff
@@ -46,16 +52,40 @@ public void setRatio(double ratio)
...
+            dutyCycleEncoder.setDistancePerRotation(distancePerRotation);
+        }
     }
 
     private Encoder initializeE4T(int channelA, int channelB)
     {
         if (channelA < 0 || channelB < 0)
         {
-            System.err.println("Encoder not activated " + channelA + " " + channelB);
+            System.err.println("E4T Encoder not activated due to negative port value" + channelA + " " + channelB);
```
Comment (dalittlet) [inline TitanRobotics2024/src/main/java/frc/robot/Subsystem/ModifiedEncoders.java:88]:
```suggestion
```

File: TitanRobotics2024/src/main/java/frc/robot/Subsystem/ModifiedEncoders.java [hunk 12]
```diff
@@ -66,7 +96,7 @@ private Encoder initializeE4T(int channelA, int channelB)
         }
         catch (Exception e)
         {
-            System.err.println("Error: Encoder Not Activated " + channelA + " " + channelB);
+            System.err.println("Error: E4T Encoder not activated due to initialization error" + channelA + " " + channelB);
```
Comment (dalittlet) [inline TitanRobotics2024/src/main/java/frc/robot/Subsystem/ModifiedEncoders.java:99]:
```suggestion
```

File: TitanRobotics2024/src/main/java/frc/robot/Subsystem/ModifiedEncoders.java [hunk 13]
```diff
@@ -76,7 +106,7 @@ private Encoder initializeQuadrature(int channelA, int channelB)
     {
         if (channelA < 0 || channelB < 0)
         {
-            System.err.println("Encoder not activated " + channelA + " " + channelB);
+            System.err.println("Quadrature Encoder not activated due to negative port value" + channelA + " " + channelB);
```
Comment (dalittlet) [inline TitanRobotics2024/src/main/java/frc/robot/Subsystem/ModifiedEncoders.java:109]:
```suggestion
```

File: TitanRobotics2024/src/main/java/frc/robot/Subsystem/ModifiedEncoders.java [hunk 14]
```diff
@@ -86,7 +116,27 @@ private Encoder initializeQuadrature(int channelA, int channelB)
         }
         catch (Exception e)
         {
-            System.err.println("Error: Encoder Not Activated " + channelA + " " + channelB);
+            System.err.println("Error: Quadrature Encoder not activated due to initialization error" + channelA + " " + channelB);
```
Comment (dalittlet) [inline TitanRobotics2024/src/main/java/frc/robot/Subsystem/ModifiedEncoders.java:119]:
```suggestion
```

File: TitanRobotics2024/src/main/java/frc/robot/Subsystem/ModifiedEncoders.java [hunk 15]
```diff
@@ -86,7 +116,27 @@ private Encoder initializeQuadrature(int channelA, int channelB)
...
+            return null;
+        }
+    }
+
+    private DutyCycleEncoder initializeDutyCycle(int portNumber, double positionOffset)
+    {
+        if (portNumber < 0)
+        {
+            System.err.println("DutyCycleEncoder not activated due to negative port value" + portNumber);
```
Comment (dalittlet) [inline TitanRobotics2024/src/main/java/frc/robot/Subsystem/ModifiedEncoders.java:128]:
```suggestion
```

File: TitanRobotics2024/src/main/java/frc/robot/Subsystem/ModifiedEncoders.java [hunk 16]
```diff
@@ -86,7 +116,27 @@ private Encoder initializeQuadrature(int channelA, int channelB)
...
+        }
+        try
+        {
+            dutyCycleEncoder = new DutyCycleEncoder(portNumber);
+            dutyCycleEncoder.setPositionOffset(positionOffset);
+            return dutyCycleEncoder;
+        }
+        catch (Exception e)
+        {
+            System.err.println("Error: DutyCycleEncoder not activated due to initialization error" + portNumber);
```
Comment (dalittlet) [inline TitanRobotics2024/src/main/java/frc/robot/Subsystem/ModifiedEncoders.java:139]:
```suggestion
```

File: TitanRobotics2024/src/main/java/frc/robot/Subsystem/ModifiedEncoders.java (hunk 10 above)
Comment (thebuilder6) [reply TitanRobotics2024/src/main/java/frc/robot/Subsystem/ModifiedEncoders.java:45]: These start up error messages are intentional, and very helpful in determining that the robot is configured properly

File: TitanRobotics2024/src/main/java/frc/robot/Subsystem/ModifiedEncoders.java (hunk 12 above)
Comment (impossiblerocket) [reply TitanRobotics2024/src/main/java/frc/robot/Subsystem/ModifiedEncoders.java:99]: only prints once

File: TitanRobotics2024/src/main/java/frc/robot/Subsystem/ModifiedEncoders.java (hunk 13 above)
Comment (impossiblerocket) [reply TitanRobotics2024/src/main/java/frc/robot/Subsystem/ModifiedEncoders.java:109]: only prints once

File: TitanRobotics2024/src/main/java/frc/robot/Subsystem/ModifiedEncoders.java (hunk 14 above)
Comment (impossiblerocket) [reply TitanRobotics2024/src/main/java/frc/robot/Subsystem/ModifiedEncoders.java:119]: only prints once

File: TitanRobotics2024/src/main/java/frc/robot/Subsystem/ModifiedEncoders.java (hunk 16 above)
Comment (impossiblerocket) [reply TitanRobotics2024/src/main/java/frc/robot/Subsystem/ModifiedEncoders.java:139]: only prints once

File: TitanRobotics2024/src/main/java/frc/robot/Subsystem/ModifiedEncoders.java (hunk 15 above)
Comment (impossiblerocket) [reply TitanRobotics2024/src/main/java/frc/robot/Subsystem/ModifiedEncoders.java:128]: only prints once

File: TitanRobotics2024/src/main/java/frc/robot/Subsystem/ModifiedEncoders.java (hunk 12 above)
Comment (impossiblerocket) [reply TitanRobotics2024/src/main/java/frc/robot/Subsystem/ModifiedEncoders.java:99]: only prints once

File: TitanRobotics2024/src/main/java/frc/robot/Subsystem/ModifiedEncoders.java (hunk 10 above)
Comment (impossiblerocket) [reply TitanRobotics2024/src/main/java/frc/robot/Subsystem/ModifiedEncoders.java:45]: only prints once

File: TitanRobotics2024/src/main/java/frc/robot/Subsystem/ModifiedEncoders.java (hunk 11 above)
Comment (impossiblerocket) [reply TitanRobotics2024/src/main/java/frc/robot/Subsystem/ModifiedEncoders.java:88]: only prints once

**Pull request merged**